create_derivatives/routines.py

```diff
@@ -129,9 +133,136 @@ def ocr_pdf(self):
 
 
 class ManifestMaker:
-    # TO DO: make manifests
+    """Creates a IIIF presentation manifest version 3 from JP2 files.
+
+    Creates manifest directory in bag's data directory and then creates manifest.
+
+    Returns:
+        A tuple containing human-readable message along with list of bag identifiers.
+        Exceptions are raised for errors along the way.
+    """
+
+    def __init__(self):
+        server_url = settings.IMAGESERVER_URL
+        self.resource_url = "{}/iiif/3/".format(server_url)
+        self.fac = ManifestFactory()
+        self.fac.set_base_prezi_uri("{}/manifests/".format(server_url))
+        self.fac.set_base_image_uri(self.resource_url)
+        self.fac.set_debug(settings.PREZI_DEBUG)
+        self.upgrader = Upgrader()
+
     def run(self):
-        pass
+        bags_with_manifests = []
+        for bag in Bag.objects.filter(process_status=Bag.PDF):
+            jp2_path = Path(bag.bag_path, "data", "JP2")
+            jp2_files = sorted([f for f in matching_files(jp2_path)])
+            manifest_dir = Path(bag.bag_path, "data", "MANIFEST")
+            if not manifest_dir.is_dir():
+                manifest_dir.mkdir()
+            self.fac.set_base_prezi_dir(str(manifest_dir))
+            self.create_manifest(jp2_files, manifest_dir, jp2_path, bag.dimes_identifier, bag.as_data)
+            bag.process_status = Bag.MANIFESTS_CREATED
+            bag.save()
+            bags_with_manifests.append(bag.dimes_identifier)
+        msg = "Manifests successfully created." if len(bags_with_manifests) else "No manifests created."
+        return msg, bags_with_manifests
+
+    def create_manifest(self, files, manifest_dir, image_dir, identifier,
```

Comment:
You may be able to reduce the number of arguments by calling for a `bag` argument, and setting some other arguments as attributes on `self` (see comment in `run` method).

Reply:
I tested with using a bag argument and setting attributes on the bag, and it kinda just resulted in the same complexity. I have instantiated some empty self attributes and set values to those during the for loop. That reduced the number of arguments I had to pass to `create_manifests`, but it's functionally the same.